static/src/javascripts/projects/common/modules/component.ts

```diff
@@ -0,0 +1,429 @@
+import { isString } from '@guardian/libs';
+import { fetchJson } from '../../../lib/fetch-json';
+
+class ComponentError {
+	constructor(message: string) {
+		return new Error(`Component: ${message}`);
+	}
+}
+
+type ManipulationType = 'append' | 'html';
+type Parent = HTMLElement | Node;
+
+const manipulate = (
+	type: ManipulationType,
+	parent: Parent,
+	elem: HTMLElement | null,
+): void => {
+	if (!elem) throw new Error('Element is null ');
+
+	switch (type) {
+		case 'append':
+			parent.appendChild(elem);
+			break;
+		case 'html':
+			if ('innerHTML' in parent) parent.innerHTML = elem.innerHTML;
+			else
+				throw new Error(
+					'Could not modify the parent’s HTML: Node type = ' +
+						String(parent.nodeType),
+				);
+			break;
+	}
+};
+
+const create = (maybeHtmlString: unknown): HTMLElement | null => {
+	if (!isString(maybeHtmlString))
+		throw new Error(
+			'response is not a valid string:' + String(maybeHtmlString),
+		);
+
+	const fragment = document.createElement('template');
+	fragment.innerHTML = maybeHtmlString;
+	const elem = fragment.content.firstElementChild?.cloneNode(true);
+
+	if (!(elem instanceof HTMLElement))
+		throw new Error('Not valid HTMLElement: ' + maybeHtmlString);
+
+	return elem;
+};
+
+class Component {
```

Comment:
This is amazing @mxdvl ! Probably one of the most difficult conversions done so far?

My only concern is with such a foundational module with a large api surface area that behaviour has been maintained - especially the event firing and listening.

I imagine the unit tests don't capture everything so its worth a thorough manual test.

Reply:
Yeah, this is on par with `fetch`. I’m not planning on releasing this before some thorough test on code.